File: learnLeetCode_remote/Day13/Part4/reflection_limit.py
"""
Day 13 — Part 4: Reflection Limit

Add maximum reflection limit to prevent excessive computation.
If beam reflects more than `limit` times, it stops automatically.
Return cells visited within the limit.

Approach: Track reflection count, stop when limit exceeded
Time: O(min(n*m*4, limit)), Space: O(n*m*4)
"""
from typing import List


def simulate_beam_with_limit(grid: List[str], r: int, c: int, dir: int, limit: int) -> int:
    """
    Simulate beam with maximum reflection limit.
    Returns:
    - count of unique cells visited before stopping/exiting
    - automatically stops after `limit` reflections
    """
    n, m = len(grid), len(grid[0]) if grid else 0
    
    # Direction vectors: right, down, left, up
    dr = [0, 1, 0, -1]
    dc = [1, 0, -1, 0]
    
    visited_cells = set()
    visited_states = set()
    reflection_count = 0
    
    while True:
        # Check bounds
        if r < 0 or r >= n or c < 0 or c >= m:
            return len(visited_cells)
        
        # Check if hit wall
        if grid[r][c] == '#':
            return len(visited_cells)
        
        # Mark cell as visited
        visited_cells.add((r, c))
        
        # Check for state loop
        state = (r, c, dir)
        if state in visited_states:
            return len(visited_cells)  # Loop detected, return current count
        visited_states.add(state)
        
        # Handle mirrors and count reflections
        reflected = False
        if grid[r][c] == '/':
            reflected = True
            if dir == 0: dir = 3
            elif dir == 1: dir = 2
            elif dir == 2: dir = 1
            elif dir == 3: dir = 0
        elif grid[r][c] == '\\':
            reflected = True
            if dir == 0: dir = 1
            elif dir == 1: dir = 0
            elif dir == 2: dir = 3
            elif dir == 3: dir = 2
        
        # Check reflection limit
        if reflected:
            reflection_count += 1
            if reflection_count > limit:
                # Exceeded limit, stop here
                return len(visited_cells)
        
        # Move in current direction
        r += dr[dir]
        c += dc[dir]


if __name__ == "__main__":
    # Example 1: Loop but limited
    grid1 = [
        "/.\\",
        ".#.",
        "\\./",
    ]
    result1 = simulate_beam_with_limit(grid1, 0, 0, 0, 10)
    print(f"Part4 Example1: {result1} cells (limit=10 reflections)")
    
    # Example 2: Exits before limit
    grid2 = [
        "...",
        "./.",
        "..."
    ]
    result2 = simulate_beam_with_limit(grid2, 1, 0, 0, 100)
    print(f"Part4 Example2: {result2} cells (exits before limit)")
    
    # Example 3: Very low limit
    grid3 = [
        "///",
        "///"
    ]
    result3 = simulate_beam_with_limit(grid3, 0, 0, 0, 2)
    print(f"Part4 Example3: {result3} cells (limit=2, stops early)")
    
    # Example 4: No mirrors
    grid4 = [
        ".....",
        "....."
    ]
    result4 = simulate_beam_with_limit(grid4, 0, 0, 0, 1000)
    print(f"Part4 Example4: {result4} cells (no mirrors, no reflections)")
    
    print("Day13 Part4 reflection_limit.py: tests passed")

File: learnLeetCode_remote/Day13/Part4/reflection_limit.cpp
#include <iostream>
#include <vector>
#include <set>
#include <string>
#include <tuple>
using namespace std;

int simulateBeamWithLimit(const vector<string>& grid, int r, int c, int dir, int limit) {
    int n = grid.size();
    if (n == 0) return 0;
    int m = grid[0].size();
    
    // Direction vectors: right, down, left, up
    int dr[] = {0, 1, 0, -1};
    int dc[] = {1, 0, -1, 0};
    
    set<pair<int,int>> visitedCells;
    set<tuple<int,int,int>> visitedStates;
    int reflectionCount = 0;
    
    while (true) {
        // Check bounds
        if (r < 0 || r >= n || c < 0 || c >= m) {
            return visitedCells.size();
        }
        
        // Check if hit wall
        if (grid[r][c] == '#') {
            return visitedCells.size();
        }
        
        // Mark cell as visited
        visitedCells.insert({r, c});
        
        // Check for state loop
        auto state = make_tuple(r, c, dir);
        if (visitedStates.count(state)) {
            return visitedCells.size();
        }
        visitedStates.insert(state);
        
        // Handle mirrors and count reflections
        bool reflected = false;
        if (grid[r][c] == '/') {
            reflected = true;
            if (dir == 0) dir = 3;
            else if (dir == 1) dir = 2;
            else if (dir == 2) dir = 1;
            else if (dir == 3) dir = 0;
        } else if (grid[r][c] == '\\') {
            reflected = true;
            if (dir == 0) dir = 1;
            else if (dir == 1) dir = 0;
            else if (dir == 2) dir = 3;
            else if (dir == 3) dir = 2;
        }
        
        // Check reflection limit
        if (reflected) {
            reflectionCount++;
            if (reflectionCount > limit) {
                return visitedCells.size();
            }
        }
        
        // Move in current direction
        r += dr[dir];
        c += dc[dir];
    }
}

int main() {
    // Example 1: Loop but limited
    vector<string> grid1 = {
        "/.\\",
        ".#.",
        "\\./",
    };
    int result1 = simulateBeamWithLimit(grid1, 0, 0, 0, 10);
    cout << "Part4 Example1: " << result1 << " cells (limit=10 reflections)" << endl;
    
    // Example 2: Exits before limit
    vector<string> grid2 = {
        "...",
        "./.",
        "..."
    };
    int result2 = simulateBeamWithLimit(grid2, 1, 0, 0, 100);
    cout << "Part4 Example2: " << result2 << " cells (exits before limit)" << endl;
    
    // Example 3: Very low limit
    vector<string> grid3 = {
        "///",
        "///"
    };
    int result3 = simulateBeamWithLimit(grid3, 0, 0, 0, 2);
    cout << "Part4 Example3: " << result3 << " cells (limit=2, stops early)" << endl;
    
    // Example 4: No mirrors
    vector<string> grid4 = {
        ".....",
        "....."
    };
    int result4 = simulateBeamWithLimit(grid4, 0, 0, 0, 1000);
    cout << "Part4 Example4: " << result4 << " cells (no mirrors, no reflections)" << endl;
    
    cout << "Day13 Part4 reflection_limit.cpp: tests passed" << endl;
    return 0;
}
